lib/common/service.js

```diff
@@ -54,8 +59,151 @@ function Service(config, options) {
   this.interceptors = [];
   this.projectId = options.projectId;
   this.projectIdRequired = config.projectIdRequired !== false;
+
+  if (config.proto) {
+    var service = config.proto.service;
+    var apiVersion = config.proto.apiVersion;
+    var rootDir = googleProtoFiles('..');
+
+    this.protoOpts = config.proto;
+    this.proto = grpc.load({
+      root: rootDir,
+      file: path.relative(rootDir, googleProtoFiles[service][apiVersion])
+    }).google[service][apiVersion];
+  }
 }
 
+Service.prototype.protoRequest = function(protoOpts, reqOpts, callback) {
+  var self = this;
+  var proto = this.proto;
+
+  if (!this.grpcCredentials) {
+    // We must establish an authClient to give to grpc.
+    this.authClient.getCredentials(function(err) {
+      if (err) {
+        callback(err);
+        return;
+      }
+
+      var authClient = self.authClient.authClient;
+
+      self.grpcCredentials = grpc.credentials.combineChannelCredentials(
+        grpc.credentials.createSsl(),
+        grpc.credentials.createFromGoogleCredential(authClient)
+      );
+
+      self.protoRequest.call(self, protoOpts, reqOpts, callback);
+    });
+    return;
+  }
+
+  var service = new proto[protoOpts.service](
+    this.protoOpts.baseUrl,
+    this.grpcCredentials
+  );
+
+  service[protoOpts.method](snakeize(reqOpts), function(err, resp) {
+    if (err) {
+      var HTTP_ERROR_CODE_MAP = {
```

Comment:
Is this something we maybe want to move to util or perhaps just outside of this function?